src/libslic3r/GCodeWriter.cpp

```diff
@@ -682,6 +682,15 @@ std::string GCodeWriter::extrude_to_xyz(const Vec3d &point, double dE, const std
     return w.string();
 }
 
+std::string GCodeWriter::retract(double retraction_length, bool toolchange)
+{
+    return this->_retract(
+        retraction_length,
+        toolchange ? m_extruder->retract_restart_extra(): m_extruder->retract_restart_extra_toolchange(),
+        "retract"
+    );
+}
+
 std::string GCodeWriter::retract(bool before_wipe)
 {
     double factor = before_wipe ? m_extruder->retract_before_wipe() : 1.;
```

Comment:
The problem there is that I don’t have the actual wipe path length or velocity available to do the calculation. Let me have a think about it, it may be a good idea!!

Reply:
take a look at here, this is the only place when we call it with the parameter set to true, we can pass the wipe length here
And same for the retract_for_toolchange function.

```
std::string GCode::retract(bool toolchange, bool is_last_retraction, LiftType lift_type)
{
    std::string gcode;

    if (m_writer.extruder() == nullptr)
        return gcode;

    // wipe (if it's enabled for this extruder and we have a stored wipe path and no-zero wipe distance)
    if (EXTRUDER_CONFIG(wipe) && m_wipe.has_path() && scale_(EXTRUDER_CONFIG(wipe_distance)) > SCALED_EPSILON) {
        gcode += toolchange ? m_writer.retract_for_toolchange(true) : m_writer.retract(true);
        gcode += m_wipe.wipe(*this, toolchange, is_last_retraction);
    }
```

Reply:
Done - moved the length calculation to a separate function so i can re use it in both the wipe logic and the retract before wipe call.

Reply:
> Done - moved the length calculation to a separate function so i can re use it in both the wipe logic and the retract before wipe call.

Nice.
The new changes are very clean now.